File: src/main/java/AoC2022_07.java
import static java.util.stream.Collectors.joining;
import static java.util.stream.Collectors.summingInt;
import static java.util.stream.Collectors.toList;

import java.util.ArrayDeque;
import java.util.Deque;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.IntStream;

import com.github.pareronia.aocd.Aocd;
import com.github.pareronia.aocd.Puzzle;

public class AoC2022_07 extends AoCBase {
    
    private static final String ROOT = "/";
    
    private final Map<String, Integer> sizes;
    
    private AoC2022_07(final List<String> input, final boolean debug) {
        super(debug);
        final Deque<String> path = new ArrayDeque<>();
        this.sizes = new HashMap<>();
        for (final String line : input.subList(1, input.size())) {
            if (line.startsWith("$ cd ")) {
                final String name = line.substring("$ cd ".length());
                if (name.equals("..")) {
                    path.removeLast();
                } else {
                    path.addLast(name);
                }
            } else if (!line.startsWith("$")) {
                final var splits = line.split(" ");
                final int size;
                if (!splits[0].equals("dir")) {
                    size = Integer.parseInt(splits[0]);
                    final var list = path.stream().collect(toList());
                    IntStream.rangeClosed(0, list.size()).forEach(i -> {
                        final String pp = IntStream.range(0, i)
                                .mapToObj(list::get)
                                .collect(joining("/"));
                        sizes.merge(ROOT + pp, size, Integer::sum);
                    });
                }
            }
        }
    }
    
    public static final AoC2022_07 create(final List<String> input) {
        return new AoC2022_07(input, false);
    }

    public static final AoC2022_07 createDebug(final List<String> input) {
        return new AoC2022_07(input, true);
    }
    
    @Override
    public Integer solvePart1() {
        return this.sizes.values().stream()
                .filter(s -> s <= 100_000)
                .collect(summingInt(Integer::valueOf));
    }

    @Override
    public Integer solvePart2() {
        final int total = this.sizes.get(ROOT);
        final int wanted = 30_000_000 - (70_000_000 - total);
        return this.sizes.values().stream()
                .filter(v -> v >= wanted)
                .sorted()
                .findFirst().orElseThrow();
    }

    public static void main(final String[] args) throws Exception {
        assert AoC2022_07.createDebug(TEST).solvePart1() == 95_437;
        assert AoC2022_07.createDebug(TEST).solvePart2() == 24_933_642;

        final Puzzle puzzle = Aocd.puzzle(2022, 7);
        final List<String> inputData = puzzle.getInputData();
        puzzle.check(
            () -> lap("Part 1", AoC2022_07.create(inputData)::solvePart1),
            () -> lap("Part 2", AoC2022_07.create(inputData)::solvePart2)
        );
    }

    private static final List<String> TEST = splitLines("""
        $ cd /
        $ ls
        dir a
        14848514 b.txt
        8504156 c.dat
        dir d
        $ cd a
        $ ls
        dir e
        29116 f
        2557 g
        62596 h.lst
        $ cd e
        $ ls
        584 i
        $ cd ..
        $ cd ..
        $ cd d
        $ ls
        4060174 j
        8033020 d.log
        5626152 d.ext
        7214296 k
        """);
}

File: src/main/python/AoC2022_07.py
#! /usr/bin/env python3
#
# Advent of Code 2022 Day 7
#


import sys
from collections import defaultdict

from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.common import aoc_samples

TEST = """\
$ cd /
$ ls
dir a
14848514 b.txt
8504156 c.dat
dir d
$ cd a
$ ls
dir e
29116 f
2557 g
62596 h.lst
$ cd e
$ ls
584 i
$ cd ..
$ cd ..
$ cd d
$ ls
4060174 j
8033020 d.log
5626152 d.ext
7214296 k
"""


Input = InputData
Output1 = int
Output2 = int


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        return input_data

    def get_sizes(self, inputs: InputData) -> dict[str, int]:
        path = list[str]()
        sizes = defaultdict[str, int](int)
        for line in tuple(inputs)[1:]:
            if line.startswith("$ cd "):
                name = line[5:]
                if name == "..":
                    path = path[:-1]
                else:
                    path.append(name)
            elif not line.startswith("$"):
                first, _ = line.split()
                if first != "dir":
                    size = int(first)
                    for i in range(len(path) + 1):
                        pp = "/".join(_ for _ in path[:i])
                        sizes["/" + pp] += size
        return sizes

    def part_1(self, inputs: Input) -> Output1:
        return sum(v for v in self.get_sizes(inputs).values() if v <= 100_000)

    def part_2(self, inputs: Input) -> Output2:
        sizes = self.get_sizes(inputs)
        wanted = 30_000_000 - (70_000_000 - sizes["/"])
        return next(v for v in sorted(sizes.values()) if v >= wanted)

    @aoc_samples(
        (
            ("part_1", TEST, 95_437),
            ("part_2", TEST, 24_933_642),
        )
    )
    def samples(self) -> None:
        pass


solution = Solution(2022, 7)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()
